### PR TITLE
raftstore: don't remove other peer's read delegate

### LINKED ISSUE #11852: Got unexpected panic during tipocket ledger test
### What version of TiKV are you using?

master (628937ab656df118b822196f407d837b4599c848)

### What operating system and CPU are you using?

kubernetes

### Steps to reproduce

run tipocket ledger test.

### What did you expect?
tikv-server shouldn't panic during the test.

### What did happened?

got unexpected panic from raftstore, the full log can be accessed [here](http://172.16.4.180:31714/artifacts-by-uid/c869dd64-ab43-49fb-a582-f6c8853508f6/tipocket-ledger-tiflash1-1641700800/output).
```
   0: tikv_util::set_panic_hook::{{closure}}
             at home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tikv/components/tikv_util/src/lib.rs:464:18
   1: std::panicking::rust_panic_with_hook
             at rustc/2faabf579323f5252329264cc53ba9ff803429a3/library/std/src/panicking.rs:626:17
   2: std::panicking::begin_panic_handler::{{closure}}
             at rustc/2faabf579323f5252329264cc53ba9ff803429a3/library/std/src/panicking.rs:517:13
   3: std::sys_common::backtrace::__rust_end_short_backtrace
             at rustc/2faabf579323f5252329264cc53ba9ff803429a3/library/std/src/sys_common/backtrace.rs:141:18
   4: rust_begin_unwind
             at rustc/2faabf579323f5252329264cc53ba9ff803429a3/library/std/src/panicking.rs:515:5
   5: core::panicking::panic_fmt
             at rustc/2faabf579323f5252329264cc53ba9ff803429a3/library/core/src/panicking.rs:92:14
   6: core::panicking::panic
             at rustc/2faabf579323f5252329264cc53ba9ff803429a3/library/core/src/panicking.rs:50:5
   7: core::option::Option<T>::unwrap
             at rustc/2faabf579323f5252329264cc53ba9ff803429a3/library/core/src/option.rs:722:21
      raftstore::store::peer::Peer<EK,ER>::post_apply
             at home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tikv/components/raftstore/src/store/peer.rs:2715:26
      raftstore::store::fsm::peer::PeerFsmDelegate<EK,ER,T>::on_apply_res
             at home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tikv/components/raftstore/src/store/fsm/peer.rs:1432:39
      raftstore::store::fsm::peer::PeerFsmDelegate<EK,ER,T>::handle_msgs
             at home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tikv/components/raftstore/src/store/fsm/peer.rs:598:21
   8: <raftstore::store::fsm::store::RaftPoller<EK,ER,T> as batch_system::batch::PollHandler<raftstore::store::fsm::peer::PeerFsm<EK,ER>,raftstore::store::fsm::store::StoreFsm<EK>>>::handle_normal
             at home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tikv/components/raftstore/src/store/fsm/store.rs:790:9
   9: batch_system::batch::Poller<N,C,Handler>::poll
             at home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tikv/components/batch-system/src/batch.rs:446:27
  10: batch_system::batch::BatchSystem<N,C>::start_poller::{{closure}}
             at home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tikv/components/batch-system/src/batch.rs:572:17
      std::sys_common::backtrace::__rust_begin_short_backtrace
             at rustc/2faabf579323f5252329264cc53ba9ff803429a3/library/std/src/sys_common/backtrace.rs:125:18
  11: std::thread::Builder::spawn_unchecked::{{closure}}::{{closure}}
             at rustc/2faabf579323f5252329264cc53ba9ff803429a3/library/std/src/thread/mod.rs:476:17
      <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at rustc/2faabf579323f5252329264cc53ba9ff803429a3/library/std/src/panic.rs:347:9
      std::panicking::try::do_call
             at rustc/2faabf579323f5252329264cc53ba9ff803429a3/library/std/src/panicking.rs:401:40
      std::panicking::try
             at rustc/2faabf579323f5252329264cc53ba9ff803429a3/library/std/src/panicking.rs:365:19
      std::panic::catch_unwind
             at rustc/2faabf579323f5252329264cc53ba9ff803429a3/library/std/src/panic.rs:434:14
      std::thread::Builder::spawn_unchecked::{{closure}}
             at rustc/2faabf579323f5252329264cc53ba9ff803429a3/library/std/src/thread/mod.rs:475:30
      core::ops::function::FnOnce::call_once{{vtable.shim}}
             at rustc/2faabf579323f5252329264cc53ba9ff803429a3/library/core/src/ops/function.rs:227:5
  12: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
             at rustc/2faabf579323f5252329264cc53ba9ff803429a3/library/alloc/src/boxed.rs:1572:9
      <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
             at rustc/2faabf579323f5252329264cc53ba9ff803429a3/library/alloc/src/boxed.rs:1572:9
      std::sys::unix::thread::Thread::new::thread_start
             at rustc/2faabf579323f5252329264cc53ba9ff803429a3/library/std/src/sys/unix/thread.rs:91:17
  13: <unknown>
  14: clone
```

### PR DESCRIPTION
**Pull request by NingLin-P**
### What is changed and how it works?

Close #11852 

What's Changed:

Don't remove read delegate besides `destroy_peer`, and for robustness, replace other `meta.readers.remove(id)` with `meta.readers.get_mut(id).makr_pending_remove`.

In `Localreader`, read delegates that mark as `pendinf_remove` will behave just like it is removed.

### Check List 

Tests 

- Integration test

Side effects

### Release note 

```release-note
raftstore: don't remove other peer's read delegate
```

### CODE REVIEW

File: tests/failpoints/cases/test_merge.rs [hunk 1]
```diff
@@ -1302,3 +1302,49 @@ fn test_merge_election_and_restart() {
     // If logs from different term are process correctly, store 2 should have latest updates.
     must_get_equal(&cluster.get_engine(2), b"k14", b"v14");
 }
+
+/// Testing that the source peer's read delegate should not be removed by the target peer
+/// and only removed when the peer is destroyed
+#[test]
+fn test_source_peer_read_delegate_after_apply() {
```
Comment (NingLin-P) [inline tests/failpoints/cases/test_merge.rs:1309]: The case describes in https://github.com/tikv/tikv/issues/11852 is hard to simulate in test, thus I test `ReadDelegate` won't be removed by target peer instead.

**Review by BusyJay — APPROVED** [review]

File: tests/failpoints/cases/test_merge.rs [hunk 2]
```diff
@@ -1302,3 +1302,49 @@ fn test_merge_election_and_restart() {
...
+    // Merge finish means the leader of the target region have call `on_ready_commit_merge`
+    pd_client.must_merge(source.get_id(), target.get_id());
+
+    // The source peer's `ReadDelegate` should not be removed yet
+    cluster.store_metas[&1]
+        .lock()
+        .unwrap()
+        .readers
+        .get(&source.get_id())
+        .unwrap();
```
Comment (5kbpers) [inline tests/failpoints/cases/test_merge.rs:1336]: Shall we ensure `pending_remove` is set to true here?

**Review by 5kbpers — APPROVED** [review]
rest LGTM

File: tests/failpoints/cases/test_merge.rs (hunk 2 above)
Comment (NingLin-P) [reply tests/failpoints/cases/test_merge.rs]: Done

**Pull request merged**